binary_search_tree/tree.py

```diff
@@ -8,51 +8,113 @@ def __init__(self, key, val = None):
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key,value)
+        else:
+            self.add_helper(self.root,key,value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        current = self.root
+        while current is not None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n) 
+    # Space Complexity: O(n) 
+    def inorder_helper(self, current,output):
+        if current:
+            self.inorder_helper(current.left,output)
+            output.append({"key":current.key, "value":current.value})
+            self.inorder_helper(current.right,output)
+        return output
+    
     def inorder(self):
-        pass
+        output = []
+        return self.inorder_helper(self.root,output)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n) 
+    # Space Complexity: O(n)  
+    def preorder_helper(self, current,output):
+        if current:
+            output.append({"key":current.key, "value":current.value})
+            self.preorder_helper(current.left,output)
+            self.preorder_helper(current.right,output)
+        return output     
     def preorder(self):
-        pass
+        output = []
+        return self.preorder_helper(self.root,output)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n) 
+    # Space Complexity: O(n) 
+    def postorder_helper(self, current,output):
+        if current:
+            self.postorder_helper(current.left,output)
+            self.postorder_helper(current.right,output)
+            output.append({"key":current.key, "value":current.value})
+        return output
+        
     def postorder(self):
-        pass
+        output = []
+        return self.postorder_helper(self.root,output)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    def height_helper(self, current):
+        if current is None:
+            return 0
+        lheight = self.height_helper(current.left)
+        rheight = self.height_helper(current.right)
+        if (lheight > rheight):
+            return lheight + 1
+        else:
+            return rheight + 1 
+            
     def height(self):
```

Comment:
✨Nice! However, time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!